AppController/lib/helperfunctions.rb

```diff
@@ -500,7 +500,16 @@ def self.setup_revision(revision_key)
     tar_dir = "#{meta_dir}/app"
     return if File.directory?(tar_dir)
 
-    Djinn.fetch_revision(revision_key)
+    # Make sure we have the applcation source. If not, we have to wait
+    # till the AC populates it.
+    app_path = "#{Djinn::PERSISTENT_MOUNT_POINT}/apps/#{revision_key}.tar.gz"
+    end_work = Time.now.to_i + SLEEP_TIME * RETRIES
+    while Time.now.to_i < end_work
+      break if File.file?(app_path)
+      Djinn.log_debug("#{app_path} is not there yet. Waiting ...")
+      Kernel.sleep(Djinn::SMALL_WAIT)
+    end
```

Comment:
This should probably handle the case when the tarball still does not exist after the timeout.